sympy/ntheory/residue_ntheory.py

```diff
@@ -1513,7 +1513,38 @@ def discrete_log(n, a, b, order=None, prime_order=None):
     """
     n, a, b = as_int(n), as_int(a), as_int(b)
     if order is None:
-        order = n_order(b, n)
+        # Compute the order and its factoring in one pass
+        # order = totient(n), factors = factorint(order)
+        factors = {}
+        for px, kx in factorint(n).items():
+            if kx > 1:
+                if px in factors:
+                    factors[px] += kx - 1
+                else:
+                    factors[px] = kx - 1
+            for py, ky in factorint(px - 1).items():
+                if py in factors:
+                    factors[py] += ky
+                else:
+                    factors[py] = ky
+        order = 1
+        for px, kx in factors.items():
+            order *= px**kx
+        # Now the `order` is the order of the group and factors = factorint(order)
+        # The order of `b` divides the order of the group.
+        order_factors = {}
+        for p, e in factors.items():
+            i = 0
+            for _ in range(e):
+                if pow(b, order // p, n) == 1:
+                   order //= p
+                   i += 1
+                else:
+                    break
+            if i<e:
+                order_factors[p] = e-i
+    else:
```

Comment:
Then, when the `order` is not `None` and `_discrete_log_pohlig_hellman` is not used, a useless calculation is performed.
It would be clearer to do the following

```python
if order is None:
    # Calculate these together
    # order = n_order(b, n)
    # prime_order = isprime(order)
    # order_factors = factorint(order)

if order < 1000: # May be calculated before checking `order`
    return _discrete_log_trial_mul(n, a, b, order)

if prime_order or isprime(order):
    if order < 1000000000000:
        return _discrete_log_shanks_steps(n, a, b, order)
    return _discrete_log_pollard_rho(n, a, b, order)
return _discrete_log_pohlig_hellman(n, a, b, order, order_factors)
```

Reply:
Good point! Fixed.